source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -180,19 +189,23 @@ bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
   ComputeBlock2HeaderMaps(structuredOrder);
   bool modified = false;
   // Add instructions with external side effects to worklist. Also add branches
-  // EXCEPT those immediately contained in an "if" selection construct.
+  // EXCEPT those immediately contained in an "if" selection construct or a loop
+  // or continue construct.
   // TODO(greg-lunarg): Handle Frexp, Modf more optimally
   call_in_func_ = false;
   func_is_entry_point_ = false;
   private_stores_.clear();
-  // Stacks to keep track of when we are inside an if-construct. When not
-  // immediately inside an in-construct,  we must assume all branches are live.
+  // Stacks to keep track of when we are inside an if- or loop-construct.
+  // When not immediately inside an if- or loop-construct,  we must assume
+  // all branches are live as we may be inside of a control construct (ie
+  // switch) which is not part of the ADCE analysis.
```

Comment:
This should not be necessary, really.  I don't see why control structures should be special wrt ADCE.  You simply need to know how to remove the whole structure when you find it dead.  What am I missing?  All this tracking makes the logic more difficult to follow.

Reply:
I can see how the reason for this can be unclear.

I am actually looking at this logic right now to fix an issue that @s-perron found with empty-if, so I will see if I can either improve the logic or the documentation in this area as part of that issue.